### PR TITLE
Feat/Add Resources (StatefulSets, DaemonSets, HPA, PodDisruptionBudgets, CronJobs, Jobs, Keda ScaledObjects)

### PR DESCRIPTION
**Pull request by samuel-esp**
## Motivation

Added the following resources, already present in the original py-kube-downscaler

- StatefulSets
- DaemonSets
- HPA
- PodDisruptionBudgets
- CronJobs
- ScaledObjects

PS: I closed the old PR, and re opened here a clean one with squashed commits and the commit message written using the convention. This commit was made with lint disabled, I'll use it when more commits made with the new lints are merged inside the main (it was preventing me to push due to errors in modules I didn't want to touch for this task)

## Changes

- Workload Interface now becomes a father interface for many sub-interfaces
- Refactored logic inside DownscaleWorkload and UpscaleWorkload, the type of scaling will be different based on the "Family" of the resource and it will be automatically understood thanks to the child interface
- Added dynamic client interface to handle future crds (Keda, Argo, ...)
- 

## Tests done

Built image and tested on a mock cluster. Will add also unit tests asap

## TODO

I kept the mod file as similar as possible to the original. I dont know if the k8s release you are targeting for is the 1.30 (most probably the answer is yes seeing the starting dependencies, even though 1.31 apis should be retrocompatible).

I did have problems with keda regarding these 2 dependencies when building the image. So we should check how to solve these compatibilities problem

- k8s.io/apimachinery v0.30.3 -> 0.31beta solves the issue
- sigs.k8s.io/controller-runtime v0.17.5 -> 0.19 solves the issue but requieres k8s.io dependencies to be at 0.31. Probabily using different golang base image different than alpine will solve the problems at 0.18.5

these problems arised because Keda releaeses under 2.15.1 are incompatible with the 0.30.3 apis. Upgrading to 0.31 and controller-runtime 0.19 solved everything and I was able to correctly test it 

- [x] I've assigned myself to this PR

### CODE REVIEW

File: internal/pkg/scalable/scaledobjects.go
Comment (jonathan-mayer) [inline internal/pkg/scalable/scaledobjects.go:57]: This func is pretty useless, and has a lot of unneeded overhead.
The functionality could be achived just with: (we don't need an error if it doesn't exist, we just care that is no longer exists after calling this)
``` go
// RemovePauseScaledObjectAnnotation will remove the keda pause annotation
func (s ScaledObject) RemovePauseScaledObjectAnnotation() {
    delete(s.Annotations, kedaPausedReplicasAnnotation)
} 
```

Possibly everything to do with this should be refactored to not have to neeed this funciton at all

File: internal/pkg/scalable/scaledobjects.go
Comment (jonathan-mayer) [inline internal/pkg/scalable/scaledobjects.go:62]: missing return

File: internal/pkg/scalable/scaledobjects.go
Comment (jonathan-mayer) [inline internal/pkg/scalable/scaledobjects.go:76]: all of these else blocks are not needed. these could all be replaced like this:
``` go
if something {
    // do something
    return anotherThing
}
// everything within else block goes here
```
try to avoid using else at all.
try to always use guard clauses.

**Review by jonathan-mayer — CHANGES_REQUESTED** [review]
I think I will leave the review at this for now. It may doesn't seem much but the requested changes should be applied anywhere with similar problems. Also i think this should be refactored. With the changes most of the funcs do almost nothing. 
Install pre commit, if it gives an error it is because of an error that should be resolved. If you want to know which error it is you can use `pre-commit run --all-files`, it will give an output of what problems exist. Also run `go test ./...` sometimes or just let pre commit do it automatically on commits.

**Comment by jonathan-mayer** [comment]
You also don't have to make an docker image, you can just run the downscaler locally with `go run ./cmd/kubedownscaler -k=/path/to/kubeconfig` (the kubeconfig has to have the permissions and the context set to the one you want to use)

File: internal/pkg/scalable/scaledobjects.go
Comment (samuel-esp) [reply internal/pkg/scalable/scaledobjects.go:57]: I didn't understand this point, do you prefer to call directly the map delete from the upscaling function without referring to this method inside the ScaledObject class?

File: internal/api/kubernetes/client.go
Comment (samuel-esp) [reply internal/api/kubernetes/client.go:100]: That's ok, I'll make both functions and scaled classes private (like right now we have for deployments)

File: internal/pkg/scalable/scaledobjects.go
Comment (samuel-esp) [reply internal/pkg/scalable/scaledobjects.go:76]: Refactored as you said, in this way. Tell me if it is ok

```
func (s ScaledObject) GetPauseScaledObjectAnnotationReplicasIfExistsAndValid() (int, bool, error) {
	if s.Annotations == nil {
		return 1, false, nil
	}
	if pausedReplicasStr, ok := s.Annotations[kedaPausedReplicasAnnotation]; ok {
		pausedReplicas, err := strconv.Atoi(pausedReplicasStr)
		if err != nil {
			return 1, false, fmt.Errorf("invalid value for annotation %s: %w", kedaPausedReplicasAnnotation, err)
		}
		return pausedReplicas, true, nil
	}

	return 1, false, nil
}
```

File: internal/pkg/scalable/scaledobjects.go
Comment (jonathan-mayer) [reply internal/pkg/scalable/scaledobjects.go:57]: well, the function would only be the delete, so i would refactor the upscaling func to allow for this behavior without needing this one-liner func.

File: internal/pkg/scalable/scalableResource.go
Comment (jonathan-mayer) [inline internal/pkg/scalable/scalableResource.go:35]:
```suggestion
// Workload is a interface for a scalable resource. It holds all shared resource specific functions
```

File: internal/api/kubernetes/client.go
Comment (samuel-esp) [reply internal/api/kubernetes/client.go:100]: completed

File: internal/pkg/scalable/scaledobjects.go
Comment (samuel-esp) [reply internal/pkg/scalable/scaledobjects.go:57]: functions deleted, used the default already implemented inside Workload Interface

File: internal/pkg/scalable/scaledobjects.go
Comment (samuel-esp) [reply internal/pkg/scalable/scaledobjects.go:62]: refactored

**Comment by jonathan-mayer** [comment]
I know this is a pretty big change, but could you refactor it like this:
```
// Workload is a interface for a scalable resource. It holds all resource specific functions
type Workload interface {
	// GetAnnotations gets the annotations of the resource
	GetAnnotations() map[string]string
	// GetNamespace gets the namespace of the resource
	GetNamespace() string
	// GetName gets the name of the resource
	GetName() string
	// GetUID gets the uid of the workload
	GetUID() types.UID
	// GetObjectKind gets the ObjectKind of the workload
	GetObjectKind() schema.ObjectKind
	// SetAnnotations sets the annotations on the resource. Changes won't be made on kubernetes until update() is called
	SetAnnotations(annotations map[string]string)
	// Update updates the resource with all changes made to it. It should only be called once on a resource
	Update(clientset *kubernetes.Clientset, dynamicClient dynamic.Interface, ctx context.Context) error
	// TODO comment
	ScaleUp()  error // error only for parsing or configuration problems
	// TODO comment
	ScaleDown(downscaleReplicas) error // error only for parsing or configuration problems
	// GetCurrentReplicas gets the current amount of replicas of the resource
	GetCurrentReplicas() (int, error)
}
```

also remove the funcs in workload that then nolonger need to be accessible.

Upscale and downscale basically do:
``` go
// DownscaleAppWorkload is the dedicated downscale function for apps/v1 workloads, excluding daemonsets
func (c client) DownscaleWorkload(replicas int, workload scalable.Workload, ctx context.Context) error {
	workload.ScaleDown()
	err = workload.Update(c.clientset, c.dynamicClient, ctx)
	if err != nil {
		return fmt.Errorf("failed to update workload: %w", err)
	}
	slog.Debug("successfully scaled down workload", "workload", workload.GetName(), "namespace", workload.GetNamespace())
	return nil
}
```

we then no longer need the other interfaces for every type, the client.RemoveOriginalReplicas, Set and Get can be private funcs within the util.go file in the scalable package

cronjobs just set the suspend
the rest just sets the replicas in whatever way it needs to eg. 
deployments sets replicas, 
hpa additionally checks for if replicas < 1
keda sets the annotation(i think)

i know this is a pretty big change but it will make this waaaaayy easier to read and understand. the scaledown dowscale replicas can be expicitly ignored with a _ in the func that dont need it like the cronjob

**Comment by samuel-esp** [comment]
Of course, just give me some time and it will be ready!

File: internal/api/kubernetes/client.go
Comment (jonathan-mayer) [inline internal/api/kubernetes/client.go:123]: oh yea i forgot about that, it was fixed in other PRs but they aren't merged yet because the other code owners were on vacation

**Review by jonathan-mayer — COMMENTED** [review]
Ill have a look over it first thing tomorrow

File: internal/api/kubernetes/client.go
Comment (samuel-esp) [reply internal/api/kubernetes/client.go:123]: That was the reason why I disabled the lint. I didn't want to modify this class before it was merged on main

File: internal/pkg/scalable/cronjobs.go
Comment (jonathan-mayer) [inline internal/pkg/scalable/cronjobs.go:5]: if the suspend is nil the cronjob is not suspended right? so why then care that it doesn't have the attribute. i would just do it like this:
```suggestion
// ScaleUp upscale the resource when the downscale period ends
func (c cronJob) ScaleUp() error {
	const suspend = false

	currentSuspendIsTrue, err := c.getSuspend()
	if err != nil {
		return fmt.Errorf("failed to get current replicas for workload: %w", err)
	}
	if !currentSuspendIsTrue {
		slog.Debug("workload is already upscaled, skipping", "workload", c.GetName(), "namespace", c.GetNamespace())
		return nil
	}

	c.setSuspend(suspend)
	return nil
}
```

File: internal/pkg/scalable/deployments.go
Comment (jonathan-mayer) [inline internal/pkg/scalable/deployments.go:58]: again use *deployment everywhere where values are directly modified

File: internal/pkg/scalable/deployments.go
Comment (jonathan-mayer) [inline internal/pkg/scalable/deployments.go:38]: I'd probably check for if the replicas is < 0 thought, as that is the lowest valid value. If you do so also change the error message below. Also new errors (errors which don't wrap another error) should be defined as a var at the top of the file (we will change those to custom errors later).

File: internal/pkg/scalable/deployments.go
Comment (jonathan-mayer) [inline internal/pkg/scalable/deployments.go:74]: I'd actually remove all the things used only for debug in all of the scaling funcs. 

```suggestion
	originalReplicas, err := getOriginalReplicas(d)
	if err != nil {
		return fmt.Errorf("failed to get original replicas for workload: %w", err)
	}
	if originalReplicas == values.Undefined {
		slog.Debug("original replicas is not set, skipping", "workload", d.GetName(), "namespace", d.GetNamespace())
		return nil
	}
```

**Review by jonathan-mayer — CHANGES_REQUESTED** [review]
Ok, I think if you should change these for all resources before I review further. I would wait with writing the tests until i say that the code is fine.

File: internal/pkg/scalable/daemonsets.go
Comment (samuel-esp) [reply internal/pkg/scalable/daemonsets.go:27]: completed

**Review by jonathan-mayer — CHANGES_REQUESTED** [review]
still have to review:
scalableResources.go
scaledobjects
statefulsets
util.go

**Comment by jonathan-mayer** [comment]
I dont know if you already did, but could you retest these changes on a cluster. (you can also wait until the review is finished)

**Review by jonathan-mayer — COMMENTED** [review]
Some small fixes, and i pushed a refactor/fix of the scaledObjects. I will still look into a way of not having that dynamic client and replacing it with a normal client with a custom scheme. You could already start writing tests for the ScaleUp and ScaleDown functions, as those shouldn't change any of the other aspects of the downscaling. It would also be awsome if you could test these resources manually in a cluster.

**Comment by samuel-esp** [comment]
I'll try to test everything asap both in the real cluster and test classes, I might be a little bit busier in the next days but I'll do my best

File: internal/pkg/scalable/scaledobjects.go
Comment (samuel-esp) [reply internal/pkg/scalable/scaledobjects.go:40]: yeah! wrong copy paste for comment

File: internal/pkg/scalable/scalableResource.go
Comment (samuel-esp) [reply internal/pkg/scalable/scalableResource.go:49]: @jonathan-mayer I thought this method was inherited from a k8s api interface, isn't it? When I tried his twin method (GetAnnotations()) in the real cluster, it worked out of the box even without actual implementations within the codebase.  If I set it to private it will require me to do a custom implementations on every resource

**Comment by samuel-esp** [comment]
@jonathan-mayer I will try to start writing tests between this evening or tomorrow, look at the unresolved topic, I have a small question on a method

File: internal/pkg/scalable/scalableResource.go
Comment (jonathan-mayer) [reply internal/pkg/scalable/scalableResource.go:49]: Oh your right, my bad.

File: internal/pkg/scalable/statefulsets.go
Comment (jonathan-mayer) [inline internal/pkg/scalable/statefulsets.go:74]: i would change resource to workload  and it would either have to be
`// ScaleDown downscales the workload` or `// ScaleDown scales down the workload`

File: internal/pkg/scalable/scalableResource.go
Comment (jonathan-mayer) [inline internal/pkg/scalable/scalableResource.go:52]: the comments for ScaleUp and ScaleDown should also be updated in the workload interface

**Review by jonathan-mayer — CHANGES_REQUESTED** [review]

**Comment by samuel-esp** [comment]
Added unit tests, unfortunately didn't have time to test in a real cluster yet

**Comment by jonathan-mayer** [comment]
LGTM I will merge after I refactored the tests a bit to have some more test cases.

File: internal/pkg/scalable/deployments.go
Comment (jonathan-mayer) [inline internal/pkg/scalable/deployments.go:72]: I just realised we need to change something back here, currently the ScaleDown with original replicas would set the orginal replicas to the downscaled replicas in the second scan being downscaled. For all resources that use orignial replicas, we need to add back the check if CurrentReplicas == DownscaleReplicas

**Review by jonathan-mayer — DISMISSED** [review]
LGTM, We should let @JTaeuber have a look over it aswell, just to make sure i didn't miss anything

File: internal/api/kubernetes/client.go
Comment (JTaeuber) [inline internal/api/kubernetes/client.go:98]:
```suggestion
		return fmt.Errorf("failed to scale down the workload: %w", err)
```

File: internal/api/kubernetes/client.go
Comment (JTaeuber) [inline internal/api/kubernetes/client.go:102]:
```suggestion
		return fmt.Errorf("failed to update the workload: %w", err)
```

File: internal/api/kubernetes/client.go
Comment (JTaeuber) [inline internal/api/kubernetes/client.go:112]:
```suggestion
		return fmt.Errorf("failed to scale up the workload: %w", err)
```

File: internal/api/kubernetes/client.go
Comment (JTaeuber) [inline internal/api/kubernetes/client.go:116]:
```suggestion
		return fmt.Errorf("failed to update the workload: %w", err)
```

**Review by JTaeuber — CHANGES_REQUESTED** [review]

**Review by JTaeuber — CHANGES_REQUESTED** [review]

File: internal/pkg/scalable/daemonsets.go
Comment (JTaeuber) [inline internal/pkg/scalable/daemonsets.go:33]:
```suggestion
// ScaleUp scales the resource up
```

File: internal/pkg/scalable/daemonsets.go
Comment (JTaeuber) [inline internal/pkg/scalable/daemonsets.go:39]:
```suggestion
// ScaleDown scales the resource down
```

**Review by JTaeuber — CHANGES_REQUESTED** [review]

File: internal/pkg/scalable/deployments.go
Comment (JTaeuber) [inline internal/pkg/scalable/deployments.go]:
```suggestion
// ScaleUp scales the resource up
```

**Review by JTaeuber — CHANGES_REQUESTED** [review]

File: internal/pkg/scalable/horizontalpodautoscalers.go
Comment (JTaeuber) [inline internal/pkg/scalable/horizontalpodautoscalers.go:13]:
```suggestion
// getHorizontalPodAutoscalers is the getResourceFunc for horizontalPodAutoscalers
```

File: internal/pkg/scalable/horizontalpodautoscalers.go
Comment (JTaeuber) [inline internal/pkg/scalable/horizontalpodautoscalers.go:16]:
```suggestion
	horizontalpodautoscalers, err := clientsets.Kubernetes.AutoscalingV2().HorizontalPodAutoscalers(namespace).List(ctx, metav1.ListOptions{TimeoutSeconds: &timeout})
```

File: internal/pkg/scalable/horizontalpodautoscalers.go
Comment (JTaeuber) [inline internal/pkg/scalable/horizontalpodautoscalers.go:20]:
```suggestion
	for _, item := range horizontalpodautoscalers.Items {
```

**Review by JTaeuber — CHANGES_REQUESTED** [review]

File: internal/pkg/scalable/jobs.go
Comment (JTaeuber) [inline internal/pkg/scalable/jobs.go:11]:
```suggestion
// getJobs is the getResourceFunc for Jobs
```

**Review by JTaeuber — CHANGES_REQUESTED** [review]

File: internal/pkg/scalable/poddisruptionbudgets.go
Comment (JTaeuber) [inline internal/pkg/scalable/poddisruptionbudgets.go:121]:
```suggestion
			return fmt.Errorf("failed to set maxUnavailable for workload: %w", err)
```

**Review by JTaeuber — CHANGES_REQUESTED** [review]

File: internal/pkg/scalable/scalableResource.go
Comment (JTaeuber) [inline internal/pkg/scalable/scalableResource.go:24]:
```suggestion
// GetResource maps the resource name to an implementation specific getResourceFunc
```

**Review by JTaeuber — CHANGES_REQUESTED** [review]

File: internal/pkg/scalable/scaledobjects.go
Comment (JTaeuber) [inline internal/pkg/scalable/scaledobjects.go:58]:
```suggestion
// ScaleUp scales the resource up
```

File: internal/pkg/scalable/scaledobjects.go
Comment (JTaeuber) [inline internal/pkg/scalable/scaledobjects.go:78]:
```suggestion
// ScaleDown scales the workload down
```
is not uniform

**Review by JTaeuber — CHANGES_REQUESTED** [review]

**Review by JTaeuber — APPROVED** [review]
LGTM

**Pull request merged**
